src/crypto/tls/handshake_server_test.go

```diff
@@ -539,6 +539,41 @@ func testCrossVersionResume(t *testing.T, version uint16) {
 	}
 }
 
+func TestTLSFlags(t *testing.T) {
+	serverConfig := &Config{
+		Certificates: []Certificate{{
+			Certificate: [][]byte{testRSACertificate},
+			PrivateKey:  testRSAPrivateKey,
+		}},
+		TLSFlagsSupported: []TLSFlag{0x50},
+	}
+	clientCert, err := X509KeyPair([]byte(clientECDSACertificatePEM), []byte(clientECDSAKeyPEM))
+	if err != nil {
+		t.Fatalf("couldn't load client certs")
+	}
+	clientConfig := &Config{
+		TLSFlagsSupported:  []TLSFlag{0x50},
+		InsecureSkipVerify: true,
+		Certificates:       []Certificate{clientCert},
+	}
+	state, _, err := testHandshake(t, clientConfig, serverConfig)
+	if err != nil {
+		t.Fatalf("handshake failed: %s", err)
+	}
+	if state.PeerTLSFlags[0] != TLSFlag(0x50) {
+		t.Fatalf("Received wrong flags")
+	}
+	if state.AgreedTLSFlags[0] != TLSFlag(0x50) {
+		t.Fatalf("Failed to agree correct flags")
+	}
+	if !state.RequestClientCert {
+		t.Fatalf("Failed to request client cert")
+	}
+	if len(state.PeerCertificates) == 0 {
+		t.Fatalf("Didn't receive correct client certs")
+	}
```

Comment:
Is `state` the client's state? If so, this test should also confirm that the server has received the flags sent by the client.

Reply:
Weirdly the order of the returned parameters is `server_state, client_state, err`.